cpp/src/arrow/c/dlpack.cc

```diff
@@ -0,0 +1,151 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/c/dlpack.h"
+
+#include "arrow/array/array_base.h"
+#include "arrow/c/dlpack_abi.h"
+#include "arrow/device.h"
+#include "arrow/type.h"
+
+namespace arrow {
+
+namespace dlpack {
+
+Status getDLDataType(const std::shared_ptr<DataType>& type, DLDataType* out) {
+  DLDataType dtype;
+  dtype.lanes = 1;
+  dtype.bits = type->bit_width();
+  switch (type->id()) {
+    case Type::INT8:
+    case Type::INT16:
+    case Type::INT32:
+    case Type::INT64:
+      dtype.code = DLDataTypeCode::kDLInt;
+      *out = dtype;
+      return Status::OK();
+    case Type::UINT8:
+    case Type::UINT16:
+    case Type::UINT32:
+    case Type::UINT64:
+      dtype.code = DLDataTypeCode::kDLUInt;
+      *out = dtype;
+      return Status::OK();
+    case Type::HALF_FLOAT:
+    case Type::FLOAT:
+    case Type::DOUBLE:
+      dtype.code = DLDataTypeCode::kDLFloat;
+      *out = dtype;
+      return Status::OK();
+    case Type::BOOL:
+      // DLPack supports byte-packed boolean values
+      return Status::TypeError("Bit-packed boolean data type not supported by DLPack.");
+    default:
+      return Status::TypeError(
+          "Can only use __dlpack__ on primitive arrays without NullType and Decimal "
+          "types.");
+  }
+}
+
+struct DLMTensorCtx {
+  std::shared_ptr<ArrayData> ref;
+  std::vector<int64_t> shape;
+  DLManagedTensor tensor;
+};
+
+static void deleter(DLManagedTensor* arg) {
+  delete static_cast<DLMTensorCtx*>(arg->manager_ctx);
+}
+
+Status ExportArray(const std::shared_ptr<Array>& arr, DLManagedTensor** out) {
+  if (arr->null_count() > 0) {
+    return Status::TypeError("Can only use __dlpack__ on arrays with no nulls.");
+  }
+
+  // Define the DLDataType struct
+  // Supported data types: int, uint, float
+  DLDataType arr_type;
+  RETURN_NOT_OK(getDLDataType(arr->type(), &arr_type));
+
+  // Create DLMTensorCtx struct with the reference to
+  // the data of the array
+  std::shared_ptr<ArrayData> array_ref = arr->data();
+  DLMTensorCtx* DLMTensor = new DLMTensorCtx;
+  DLMTensor->ref = array_ref;
+
+  // Define DLManagedTensor struct defined by
+  // DLPack (dlpack_structure.h)
+  DLManagedTensor* dlm_tensor = &DLMTensor->tensor;
+  dlm_tensor->manager_ctx = DLMTensor;
+  dlm_tensor->deleter = &deleter;
+
+  // Define the data pointer to the DLTensor
+  // If array is of length 0, data pointer should be NULL
```

Comment:
Is this specified somewhere?

Reply:
Yes, in the DLPack Python spec just above the Warning at the end of the page: https://dmlc.github.io/dlpack/latest/python_spec.html#syntax-for-data-interchange-with-dlpack

![Screenshot 2023-12-05 at 14 32 41](https://github.com/apache/arrow/assets/16418547/ba56d1fa-2d53-447e-aea8-4c6340fe7682)

Reply:
Hmm, that is weird. @kkraus14 Do you know the rationale for this? Allowing the data to be NULL for 0-sized arrays is one thing, but mandating that it is NULL seems silly.

Reply:
I do not know the rationale for this unfortunately.

@rgommers do you happen to know the rationale for this?

Reply:
It doesn't mandate that it must be NULL, but rather that it must be "NULL or 0". That seems reasonable to me, even though in theory any values for strides works. It is reasonable because there is enough evidence that in practice users regularly make mistakes with these strides - which is why for a long time NumPy had a "relaxed strides" build option to put garbage values in the strides so mistakes with them could be caught.

These Python API docs here simply transcribed what was already contained in a code comment: https://github.com/dmlc/dlpack/blame/main/include/dlpack/dlpack.h#L227. This seems to have been the requirement in DLPack from the start, see https://github.com/dmlc/dlpack/issues/35.

Reply:
Note that it is about the data pointer here, while your link to the dlpack code points to the strides.

Reply:
@rgommers

> It doesn't mandate that it must be NULL, but rather that it must be "NULL or 0".

0 in a pointer context means NULL to me, so I'm not sure which distinction you're drawing here. I might miss a subtlety, but the spec should probably be clearer if it doesn't mean "must be NULL".

> even though in theory any values for strides works

The highlighted text talks about _the data pointer_, not about strides.

Reply:
Argh sorry, it's early. Did a bit more digging, and there is no requirement documented in `dlpack.h`, nor a review comment on those Python docs when they were added. The one single mention of the data pointer being NULL I can find is [this comment](https://github.com/dmlc/dlpack/issues/51#issuecomment-849829016) by @leofang. Leo, do you have any thoughts on this?

Reply:
Well, it looks like [this PR](https://github.com/dmlc/dlpack/pull/93) was quickly approved :-D

Reply:
That's because most of it wasn't new content, but extensively reviewed over several PRs, including by the lead DLPack maintainer. 

Digging into the largest of those PRs did help me find the relevant data pointer comment: https://github.com/data-apis/array-api/pull/106#discussion_r557091538: 
"_it'd be also nice to note if the array is of zero size, in the DLTensor its data pointer should set to nullptr (or 0). In the events where strides are not specified, this is the only effective way to inform the consumer that a zero-size array is coming. Numba/CuPy received a few bug reports when CAI v0/v1 were not clear about this, thus leading to the v2 revision._"

I'm not sure that the above is actually true, because there is a `shape` field as well. But checking `data` for being NULL is perhaps easier.

Reply:
I think the convention came from the old days of CUDA Array Interface where either `None` or `0` could be used, but I guess it doesn't really matter to a C interface. Indeed, `shape` can be used to check whether an array is 0-size, but why bother asking the consumer to do an extra check when the producer already has this knowledge? 

From the ergonomics perspective, we know the pointer to a 0-size array is meaningless and shouldn't be used, and all libraries handle this case differently (ex: NumPy would return a pointer address that likely comes from its private memory reservation, while CuPy returns null), so setting it to 0 universally was the safest thing to align all libraries and tell the consumer "do not use the pointer."

Reply:
Indeed. I will send a PR to DLPAck repo later. Good suggestion!

Reply:
xref: https://github.com/dmlc/dlpack/pull/133